test/e2e/test_helm_install_upgrade.go

```diff
@@ -0,0 +1,403 @@
+						GenerateName: "gw-upgrade-120-123-",
+						Labels: map[string]string{
+							"gw-upgrade-120-123": "true",
+						},
+					},
+					Spec: gatewayv1.GatewaySpec{
+						GatewayClassName: gatewayv1.ObjectName("gwclass-upgrade-120-123"),
+						Listeners: []gatewayv1.Listener{{
+							Name:     "http",
+							Protocol: gatewayv1.HTTPProtocolType,
+							Port:     gatewayv1.PortNumber(80),
+						}},
+					},
+				},
+			},
+			assertions: []assertion{
+				{
+					Name: "gateway is programmed",
+					Func: func(c *assert.CollectT, cl *testutils.K8sClients) {
+						gws, err := cl.GatewayClient.GatewayV1().Gateways(e.Namespace.Name).List(ctx, metav1.ListOptions{
+							LabelSelector: "gw-upgrade-120-123=true",
+						})
+						require.NoError(c, err)
+						require.Len(c, gws.Items, 1)
+						gw := &gws.Items[0]
+						assert.True(c, gateway.IsProgrammed(gw))
+						assert.True(c, gateway.AreListenersProgrammed(gw))
+					},
+				},
+			},
+		},
+		{
+			name:             "upgrade from 1.2.3 to current",
+			fromVersion:      "1.2.3",
+			upgradeToCurrent: true,
+			objectsToDeploy: []client.Object{
+				&operatorv1beta1.GatewayConfiguration{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "gwconf-upgrade-123-current",
+					},
+					Spec: baseGatewayConfigurationSpec(),
+				},
+				&gatewayv1.GatewayClass{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "gwclass-upgrade-123-current",
+					},
+					Spec: gatewayv1.GatewayClassSpec{
+						ParametersRef: &gatewayv1.ParametersReference{
+							Group:     gatewayv1.Group(operatorv1beta1.SchemeGroupVersion.Group),
+							Kind:      gatewayv1.Kind("GatewayConfiguration"),
+							Namespace: (*gatewayv1.Namespace)(&e.Namespace.Name),
+							Name:      "gwconf-upgrade-123-current",
+						},
+						ControllerName: gatewayv1.GatewayController(vars.ControllerName()),
+					},
+				},
+				&gatewayv1.Gateway{
+					ObjectMeta: metav1.ObjectMeta{
+						GenerateName: "gw-upgrade-123-current-",
+						Labels: map[string]string{
+							"gw-upgrade-123-current": "true",
+						},
+					},
+					Spec: gatewayv1.GatewaySpec{
+						GatewayClassName: gatewayv1.ObjectName("gwclass-upgrade-123-current"),
+						Listeners: []gatewayv1.Listener{{
+							Name:     "http",
+							Protocol: gatewayv1.HTTPProtocolType,
+							Port:     gatewayv1.PortNumber(80),
+						}},
+					},
+				},
+			},
+			assertions: []assertion{
+				{
+					Name: "gateway is programmed",
+					Func: func(c *assert.CollectT, cl *testutils.K8sClients) {
+						gws, err := cl.GatewayClient.GatewayV1().Gateways(e.Namespace.Name).List(ctx, metav1.ListOptions{
+							LabelSelector: "gw-upgrade-123-current=true",
+						})
+						require.NoError(c, err)
+						require.Len(c, gws.Items, 1)
+						gw := &gws.Items[0]
+						assert.True(c, gateway.IsProgrammed(gw))
+						assert.True(c, gateway.AreListenersProgrammed(gw))
+					},
+				},
+			},
+		},
+		{
+			name:             "upgrade from nightly to current",
```

Comment:
Does it make sense to test this upgrade path? I believe this is not an upgrade users would perform - maybe it's better to save CI time, WDYT?

Reply:
Tempting to get shorter CI runs 🙃 but this can potentially help us find subtle breaking changes in our most recent commits (that e.g. are not included in the latest tag).